### DISCUSSION TITLE
Building from source

### DISCUSSION THREAD

**Discussion by samizdatco** [plotdevice/plotdevice#59] [Announcements]
If you're interested in using the bleeding edge code from the repository, there are a couple of ways to build it yourself.

### Module builds

Your best bet is to create a virtual environment and install the module into it. Something like this will work:

```console
git clone https://github.com/plotdevice/plotdevice.git
python3 -m venv env
./env/bin/pip3 install -U pip wheel
./env/bin/pip3 install ./plotdevice
```

You can then either use the command line tool like this:
```console
./env/bin/plotdevice plotdevice/examples/Drawing/Logo.pv 
```

or use the virtual environment's `python3` and import it yourself:
```console
❯ ./env/bin/python3
Python 3.8.9 (default, May 17 2022, 12:55:41) 
[Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from plotdevice import *
```

### Application builds

If you want the full application, and you have Xcode installed, the easiest approach is to cd into the repository and type:
```console
python3 setup.py app
```

You can also use `py2app` but it requires you to install a local copy of PyObjC first (which the `dev` setup command will add to the `deps/local` folder for you):
```console
python3 setup.py dev
./deps/local/3.8.9/bin/python3 setup.py py2app
```
> Note that the `3.8.9` in that second command depends on the version of python3 you used for the first command. Look inside the `deps/local` folder to see what's there. 

Once the `app` or `py2app` build completes, you'll find the application in the `dist` directory:
```console
open -R dist/PlotDevice.app
```

**Comment by dimitre** [comment]
Thanks for the guide! I'm updating old Nodebox and Plotdevice projects with the latest!

**Comment by dimitre** [comment]
I've noticed the app built in this process have its own copy of python inside the application.
Is it possible to run using system python3 ? this way I can use fontforge module too :)
